File: apps/vsc-plugin/src/services/agent-bridge.service.ts
import { OutputChannel } from 'vscode';
import { WebSocketService, WebSocketListener } from './websocket.service';

export interface AgentState {
  status: string;
  currentTaskId: string | null;
  currentTaskDescription: string | null;
  progress: number;
  model: string;
}

export class AgentBridgeService implements WebSocketListener {
  private static instance: AgentBridgeService;
  private log: OutputChannel;

  agentState: AgentState = {
    status: 'idle',
    currentTaskId: null,
    currentTaskDescription: null,
    progress: 0,
    model: 'unknown',
  };

  private constructor(log: OutputChannel) {
    this.log = log;
    WebSocketService.getInstance().addListener(this);
  }

  static initialize(log: OutputChannel): AgentBridgeService {
    AgentBridgeService.instance = new AgentBridgeService(log);
    return AgentBridgeService.instance;
  }

  static getInstance(): AgentBridgeService {
    if (!AgentBridgeService.instance) {
      throw new Error('AgentBridgeService not initialized');
    }
    return AgentBridgeService.instance;
  }

  onConnected(): void {
    this.log.appendLine('Agent bridge connected');
    this.broadcastAgentState();
  }

  onDisconnected(reason: string): void {
    this.log.appendLine(`Agent bridge disconnected: ${reason}`);
  }

  onMessage(type: string, payload: Record<string, unknown>): void {
    switch (type) {
      case 'agent_command':
        this.handleAgentCommand(payload);
        break;
      case 'session_update':
        this.log.appendLine('Session update received');
        break;
    }
  }

  onError(error: string): void {
    this.log.appendLine(`Agent bridge error: ${error}`);
  }

  private handleAgentCommand(payload: Record<string, unknown>): void {
    const commandType = payload.type as string;
    if (!commandType) { return; }
    this.log.appendLine(`Received agent command: ${commandType}`);

    switch (commandType) {
      case 'start_task': {
        const inner = payload.payload as Record<string, unknown> | undefined;
        const prompt = inner?.prompt as string;
        if (prompt) { this.startTask(prompt); }
        break;
      }
      case 'stop_task':
        this.stopCurrentTask();
        break;
      case 'approve_action':
        this.approveCurrentAction();
        break;
      case 'reject_action':
        this.rejectCurrentAction();
        break;
      case 'provide_input': {
        const inner = payload.payload as Record<string, unknown> | undefined;
        const input = inner?.input as string;
        if (input) { this.provideInput(input); }
        break;
      }
      case 'cancel_task':
        this.cancelCurrentTask();
        break;
    }
  }

  startTask(prompt: string): void {
    this.agentState = {
      ...this.agentState,
      status: 'running',
      currentTaskDescription: prompt,
      progress: 0,
    };
    this.broadcastAgentState();
    this.log.appendLine(`Starting task: ${prompt}`);
  }

  stopCurrentTask(): void {
    this.agentState = { ...this.agentState, status: 'idle', progress: 0 };
    this.broadcastAgentState();
    this.log.appendLine('Task stopped');
  }

  approveCurrentAction(): void {
    this.log.appendLine('Action approved');
    this.broadcastEvent('action_approved', {});
  }

  rejectCurrentAction(): void {
    this.log.appendLine('Action rejected');
    this.broadcastEvent('action_rejected', {});
  }

  provideInput(input: string): void {
    this.log.appendLine(`Input provided: ${input.substring(0, 50)}...`);
    this.broadcastEvent('input_provided', { input });
  }

  cancelCurrentTask(): void {
    this.agentState = { ...this.agentState, status: 'idle', currentTaskId: null, progress: 0 };
    this.broadcastAgentState();
    this.log.appendLine('Task cancelled');
  }

  updateAgentStatus(status: string, taskDescription?: string, progress = 0): void {
    this.agentState = {
      ...this.agentState,
      status,
      currentTaskDescription: taskDescription ?? this.agentState.currentTaskDescription,
      progress,
    };
    this.broadcastAgentState();
  }

  private broadcastAgentState(): void {
    const ws = WebSocketService.getInstance();
    if (!ws.isConnected) { return; }
    ws.sendAgentEvent('status_changed', this.agentState.currentTaskId || '', {
      status: this.agentState.status,
      currentTaskDescription: this.agentState.currentTaskDescription,
      progress: this.agentState.progress,
      model: this.agentState.model,
    });
  }

  private broadcastEvent(eventType: string, data: Record<string, unknown>): void {
    const ws = WebSocketService.getInstance();
    if (!ws.isConnected) { return; }
    ws.sendAgentEvent(eventType, this.agentState.currentTaskId || '', data);
  }
}

File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/AgentBridgeService.kt
package com.windsurf.controller.services

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import java.io.File
import java.util.Base64

@Service(Service.Level.APP)
class AgentBridgeService : WebSocketService.WebSocketListener {

    private val logger = Logger.getInstance(AgentBridgeService::class.java)
    private val gson = Gson()

    data class AgentState(
        var status: String = "idle",
        var currentTaskId: String? = null,
        var currentTaskDescription: String? = null,
        var progress: Int = 0,
        var model: String = "unknown"
    )

    var agentState = AgentState()
        private set

    init {
        WebSocketService.getInstance().addListener(this)
    }

    override fun onConnected() {
        logger.info("Agent bridge connected")
        broadcastAgentState()
    }

    override fun onDisconnected(reason: String) {
        logger.info("Agent bridge disconnected: $reason")
    }

    override fun onMessage(type: String, payload: JsonObject) {
        when (type) {
            "agent_command" -> handleAgentCommand(payload)
            "session_update" -> handleSessionUpdate(payload)
        }
    }

    override fun onError(error: String) {
        logger.warn("Agent bridge error: $error")
    }

    private fun handleAgentCommand(payload: JsonObject) {
        val commandType = payload.get("type")?.asString ?: return
        logger.info("Received agent command: $commandType")

        val inner = payload.getAsJsonObject("payload")

        when (commandType) {
            "start_task" -> {
                var prompt = inner?.get("prompt")?.asString ?: return
                // Handle file attachments
                val files = inner.getAsJsonArray("files")
                if (files != null && files.size() > 0) {
                    for (el in files) {
                        val f = el.asJsonObject
                        val filename = f.get("filename")?.asString ?: continue
                        val base64 = f.get("base64")?.asString ?: continue
                        val tmp = File(System.getProperty("java.io.tmpdir"), "codeagent-${System.currentTimeMillis()}-$filename")
                        tmp.writeBytes(Base64.getDecoder().decode(base64))
                        prompt = "@${tmp.absolutePath} $prompt"
                        // Clean up after 2 min
                        Thread { Thread.sleep(120_000); tmp.delete() }.start()
                    }
                }
                startTask(prompt)
            }
            "stop_task" -> stopCurrentTask()
            "approve_action" -> approveCurrentAction()
            "reject_action" -> rejectCurrentAction()
            "provide_input" -> {
                val input = inner?.get("input")?.asString ?: return
                provideInput(input)
            }
            "cancel_task" -> cancelCurrentTask()
            "select_option" -> {
                val targetIndex = inner?.get("index")?.asInt ?: 0
                val currentIndex = inner?.get("currentIndex")?.asInt ?: 0
                val terminal = TerminalAgentService.getInstance()
                terminal.selectOption(targetIndex, currentIndex)
            }
            "escape_key" -> {
                TerminalAgentService.getInstance().sendEscape()
            }
            "get_context" -> {
                // Not available via IDE plugin — return minimal response
                logger.info("get_context not supported in IDE plugin")
            }
            "resume_session" -> {
                val sessionId = inner?.get("id")?.asString ?: return
                val auto = inner.get("auto")?.asBoolean ?: false
                val resumePrompt = if (auto) "--resume $sessionId --dangerously-skip-permissions" else "--resume $sessionId"
                val terminal = TerminalAgentService.getInstance()
                terminal.sendRawToTerminal("\u0003") // Ctrl+C
                Thread.sleep(500)
                terminal.sendPromptToClaudeCode(resumePrompt)
            }
        }
    }

    private fun handleSessionUpdate(payload: JsonObject) {
        logger.info("Session update received")
    }

    fun startTask(prompt: String) {
        agentState = agentState.copy(
            status = "running",
            currentTaskDescription = prompt,
            progress = 0
        )
        broadcastAgentState()

        // TODO: Integrate with actual Windsurf agent API
        logger.info("Starting task: $prompt")
    }

    fun stopCurrentTask() {
        agentState = agentState.copy(status = "idle", progress = 0)
        broadcastAgentState()
        logger.info("Task stopped")
    }

    fun approveCurrentAction() {
        logger.info("Action approved")
        broadcastEvent("action_approved", JsonObject())
    }

    fun rejectCurrentAction() {
        logger.info("Action rejected")
        broadcastEvent("action_rejected", JsonObject())
    }

    fun provideInput(input: String) {
        logger.info("Input provided: ${input.take(50)}...")
        broadcastEvent("input_provided", JsonObject().apply {
            addProperty("input", input)
        })
    }

    fun cancelCurrentTask() {
        agentState = agentState.copy(status = "idle", currentTaskId = null, progress = 0)
        broadcastAgentState()
        logger.info("Task cancelled")
    }

    fun updateAgentStatus(status: String, taskDescription: String? = null, progress: Int = 0) {
        agentState = agentState.copy(
            status = status,
            currentTaskDescription = taskDescription ?: agentState.currentTaskDescription,
            progress = progress
        )
        broadcastAgentState()
    }

    private fun broadcastAgentState() {
        val ws = WebSocketService.getInstance()
        if (!ws.isConnected) return

        val data = JsonObject().apply {
            addProperty("status", agentState.status)
            addProperty("currentTaskDescription", agentState.currentTaskDescription)
            addProperty("progress", agentState.progress)
            addProperty("model", agentState.model)
        }
        ws.sendAgentEvent("status_changed", agentState.currentTaskId ?: "", data)
    }

    private fun broadcastEvent(eventType: String, data: JsonObject) {
        val ws = WebSocketService.getInstance()
        if (!ws.isConnected) return
        ws.sendAgentEvent(eventType, agentState.currentTaskId ?: "", data)
    }

    companion object {
        fun getInstance(): AgentBridgeService =
            ApplicationManager.getApplication().getService(AgentBridgeService::class.java)
    }
}
